File: app/src/main/java/ru/snowadv/civic_climate_control/DevicesManager.java
package ru.snowadv.civic_climate_control;

import android.app.Activity;
import android.content.Context;
import android.hardware.usb.UsbManager;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public enum DevicesManager {
    INSTANCE;

    private static final String TAG = "DevicesManager";

    public Map<String, String> getDevicesList(Context context) {
        return getDevicesList(getUsbManager(context));
    }

    public Map<String, String> getDevicesList(UsbManager usbManager) {
        if(usbManager == null) {
            return new HashMap<>();
        }
        return usbManager.getDeviceList().values().stream()
                .map(SerializableUsbDevice::new)
                .collect(Collectors.toMap(SerializableUsbDevice::toString,
                        SerializableUsbDevice::toJson));
    }



    private UsbManager getUsbManager(Context context) {
        return context == null ? null : (UsbManager) context.getSystemService(Context.USB_SERVICE);
    }



}

File: app/src/main/java/ru/snowadv/civic_climate_control/DevicesManager.kt
package ru.snowadv.civic_climate_control

import android.Manifest
import android.bluetooth.BluetoothManager
import android.content.Context
import android.content.pm.PackageManager
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbManager
import android.os.Build
import androidx.core.app.ActivityCompat
import java.util.stream.Collectors

enum class DevicesManager {
    INSTANCE;

    private val TAG = "DevicesManager"
    fun getUsbDevices(context: Context?): Map<String, String> = getUsbDevices(getUsbManager(context))

    fun getUsbDevices(usbManager: UsbManager?): Map<String, String> {
        return usbManager?.deviceList?.values?.stream()
            ?.map { usbDevice: UsbDevice? -> SerializableUsbDevice(usbDevice) }
            ?.collect(
                Collectors.toMap(
                    { obj: SerializableUsbDevice -> obj.toString() },
                    { obj: SerializableUsbDevice -> obj.toJson() })
            ) ?: HashMap()
    }
    fun checkBtPermission(context: Context): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.BLUETOOTH_CONNECT
            ) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }
    }
    fun getBluetoothDevices(context: Context, noPermissionCallback: () -> Unit): Map<String, String> {
        if (checkBtPermission(context)) {
            (context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter
                ?.bondedDevices?.associateBy ({it.address}, {"${it.name} [${it.address}]"})?.let { return it }
        } else {
            noPermissionCallback()
        }
        return HashMap()
    }


    private fun getUsbManager(context: Context?): UsbManager? {
        return context?.getSystemService(Context.USB_SERVICE) as UsbManager?
    }

}